inductiva/fluids/scenarios/coastal_area/random_bathymetry_utils.py

```diff
@@ -0,0 +1,251 @@
+
+    # Make sure we are only getting indexes inside the map.
+    if enforce_inner_indexes:
+        all_positive = (src_idxs[:, 0] >= 0) & (src_idxs[:, 1] >= 0)
+        max_x, max_y = bathymetry_array.shape
+        all_bounded = (src_idxs[:, 0] < max_x) & (src_idxs[:, 1] < max_y)
+        # Filter out to make sure all indices are inside the map.
+        src_idxs = src_idxs[all_positive & all_bounded]
+
+    # Fetch elements by list of x and list of y coords and return the mean.
+    elements = bathymetry_array[src_idxs[:, 0], src_idxs[:, 1]]
+    return np.mean(elements)
+
+
+def diamond_step(bathymetry_array, inter_center_distance, noise_range):
+    """The diamond step performed at a given inter center distance on a map.
+                            +-----------+
+                            | D |   | D |
+                            |---+---+---|
+                            |   | X |   |
+                            |---+---+---|
+                            | D |   | D |
+                            +-----------+
+    For each square in the array, set the midpoint to be the average of the four
+    corner points plus a random value."""
+
+    diamond_offsets = [[-1, -1], [-1, 1], [1, 1], [1, -1]]
+    map_length = bathymetry_array.shape[0]
+    kernel_scale = inter_center_distance // 2
+
+    for i in range(kernel_scale, map_length, inter_center_distance):
+        for j in range(kernel_scale, map_length, inter_center_distance):
+            bathymetry_array[i,
+                             j] = kernel_avg(diamond_offsets, bathymetry_array,
+                                             i, j, kernel_scale)
+            bathymetry_array[i, j] += random.uniform(-noise_range, noise_range)
+
+
+def square_step(bathymetry_array, inter_center_distance, noise_range):
+    """The square step performed at a given inter center distance on a map.
+                            +-----------+
+                            |   | S |   |
+                            |---+---+---|
+                            | S | X | S |
+                            |---+---+---|
+                            |   | S |   |
+                            +-----------+
+    For each square in the array, set the midpoint to be the average of the four
+    corner points plus a random value."""
+
+    square_offsets = [[-1, 0], [0, -1], [1, 0], [0, 1]]
+    map_length = bathymetry_array.shape[0]
+    kernel_scale = inter_center_distance // 2
+
+    # Square Step, rows
+    for i in range(kernel_scale, map_length, inter_center_distance):
+        for j in range(0, map_length, inter_center_distance):
+            bathymetry_array[i,
+                             j] = kernel_avg(square_offsets, bathymetry_array,
+                                             i, j, kernel_scale)
+            bathymetry_array[i, j] += random.uniform(-noise_range, noise_range)
+
+    # Square Step, cols
+    for i in range(0, map_length, inter_center_distance):
```

Comment:
All of this double four loops could be transformed into a function, where the main change is in the for loop range.

Reply:
This would also point to the main differences between steps of the algorithm.